src/kernels/MIOpenPdist.cpp

```diff
@@ -0,0 +1,220 @@
+                                         tensor_view_t<2> input_tv,
+                                         tensor_view_t<1> output_tv,
+                                         tensor_view_t<1> grad_tv)
+{
+    pdist_backward<INPUT_TYPE>(
+        input, output, grad, input_grad, p, n2, n2_squared_minus_1, input_tv, output_tv, grad_tv);
+}
+
+template <typename DTYPE>
+__device__ void pdist_backward_contiguous(const DTYPE* __restrict__ input,
+                                          const DTYPE* __restrict__ output,
+                                          const DTYPE* __restrict__ grad, // output_grad
+                                          DTYPE* __restrict__ input_grad,
+                                          double p,
+                                          double n2,
+                                          double n2_squared_minus_1,
+                                          long N,
+                                          long NO,
+                                          long M)
+
+{
+    // NO = N(N-1)/2
+    // gws = {NO * M}
+    // input = {N, M}
+    // output = {NO}
+    // grad = {NO}
+    // input_grad = {N - 1, N, M} = {NO * 2, M}
+
+    const uint64_t gid = blockIdx.x * blockDim.x + threadIdx.x;
+
+    uint64_t k = gid / M;
+    uint64_t m = gid % M;
+
+    if(k >= NO)
+        return;
+
+    // pair(row_i, row_j) corresponding to output[k]
+    // e.g. pair(point_A, point_B)
+    long i = n2 - sqrt(n2_squared_minus_1 - 2 * k);
+    long j = k - N * i + i * (i + 1) / 2 + i + 1;
+
+    // Pair of gradients corresponding to pair(i,j)
+    // e.g. pair(input_grad(A,B), input_grad(B,A))
+    // those 2 elements in pair have opposite signs
+    // just similar to dist(A,B) vs. dist(B,A)
+    long ib = j - i - 1;
+    long jb = N - 2 - i;
+
+    FLOAT_ACCUM grad_k   = CVT_FLOAT2ACCUM(grad[k]);
+    FLOAT_ACCUM output_k = CVT_FLOAT2ACCUM(output[k]);
+
+    FLOAT_ACCUM diff = CVT_FLOAT2ACCUM(input[i * M + m]) - CVT_FLOAT2ACCUM(input[j * M + m]);
+
+    FLOAT_ACCUM p_ = static_cast<FLOAT_ACCUM>(p);
+
+    FLOAT_ACCUM res = backward(diff, grad_k, output_k, p_);
+
+    input_grad[ib * N * M + i * M + m] = CVT_ACCUM2FLOAT(res);  // dist(i,j)
+    input_grad[jb * N * M + j * M + m] = CVT_ACCUM2FLOAT(-res); // dist(j,i)
+}
+
+extern "C" __global__ void
+PdistBackwardContiguous(const INPUT_TYPE* __restrict__ input,
+                        const INPUT_TYPE* __restrict__ output,
+                        const INPUT_TYPE* __restrict__ grad, // output_grad
+                        INPUT_TYPE* __restrict__ input_grad,
+                        double p,
+                        double n2,
+                        double n2_squared_minus_1,
+                        long N,
+                        long NO,
+                        long M)
+{
+    pdist_backward_contiguous<INPUT_TYPE>(
+        input, output, grad, input_grad, p, n2, n2_squared_minus_1, N, NO, M);
+}
```

Comment:
I find CL code do not use this contiguous func and this make no difference with the above function because you use contiguous indexing input_grad[ib * N * M + i * M + m] in the above function

Reply:
TODO: Fix non-cont implementation